### PR TITLE
Global SequenceDictionary from Reference

### PR DESCRIPTION
**Pull request by erictu**
Changelog:
-SequenceDictionary is now set only from the reference file (fasta, adam) 
-RDD caching for referenceRDD (for #82 and #84) and for intervalRDDs
-adds idea to gitignore to solve #103

### CODE REVIEW

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 1]
```diff
@@ -258,8 +242,10 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
           val data = loadFromFile(reg, k)
           if (intRDD == null) {
             intRDD = IntervalRDD(data)
+            intRDD.persist(StorageLevel.MEMORY_AND_DISK)
           } else {
             intRDD = intRDD.multiput(data)
+            intRDD.persist(StorageLevel.MEMORY_AND_DISK)
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: Are all these persist calls for test? They probably shouldnt be pushed to master

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 2]
```diff
@@ -59,39 +59,23 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
...
+  def setDictionary(seqDict: SequenceDictionary) {
+    dict = seqDict
   }
 
   // Stores location of sample at a given filepath
-  def loadSample(sampleId: String, filePath: String) {
+  def loadSample(sampleId: String, filePath: String, seqdict: SequenceDictionary) {
     if (dict == null) {
-      setDictionary(filePath)
+      setDictionary(seqdict)
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: this if (dict == null) statement should be removed.

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 3]
```diff
@@ -59,39 +59,23 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
...
-      // val recs: RDD[SequenceRecord] = projected.map(rec => SequenceRecord(rec.getVariant.getContig.getContigName, (rec.getVariant.end - rec.getVariant.start)))
-      // dict = recs.aggregate(SequenceDictionary())(
-      //   (dict: SequenceDictionary, rec: SequenceRecord) => dict + rec,
-      //   (dict1: SequenceDictionary, dict2: SequenceDictionary) => dict1 ++ dict2)
-      dict = new SequenceDictionary(Vector(SequenceRecord("20", 25000000L),
-        SequenceRecord("chrM", 2000L),
-        SequenceRecord("chr3", 2000L)))
-    }
+  def setDictionary(seqDict: SequenceDictionary) {
+    dict = seqDict
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: This function isnt necessary. The dict can now just be set upon Lazy initialization

File: mango-core/src/test/scala/org/bdgenomics/mango/models/LazyMaterializationSuite.scala [hunk 4]
```diff
@@ -102,7 +102,7 @@ class LazyMaterializationSuite extends ADAMFunSuite {
 
     var lazyMat = LazyMaterialization[AlignmentRecord](sc, 10)
     val bigRegion = new ReferenceRegion("M", 0L, 20000L)
-    lazyMat.loadSample(sample1, bamFile)
+    lazyMat.loadSample(sample1, bamFile, sd)
```
Comment (akmorrow13) [inline mango-core/src/test/scala/org/bdgenomics/mango/models/LazyMaterializationSuite.scala]: loadSample shouldnt require and sd. this should be set on initialization

File: mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala [hunk 5]
```diff
@@ -139,9 +150,8 @@ object VizReads extends BDGCommandCompanion with Logging {
       case Some(_) => {
         val end: Long = Math.min(region.end, seqRecord.get.length)
         if (VizReads.referencePath.endsWith(".adam")) {
-          val pred: FilterPredicate = (LongColumn("fragmentStartPosition") >= region.start) && (LongColumn("fragmentStartPosition") <= region.end)
-          val referenceRDD: RDD[NucleotideContigFragment] = VizReads.sc.loadParquetContigFragments(VizReads.referencePath, predicate = Some(pred))
-          Option(referenceRDD.adamGetReferenceString(region))
+          refRDD.persist(StorageLevel.MEMORY_AND_DISK)
```
Comment (akmorrow13) [inline mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala]: again, why are we calling all these persists?

Comment (erictu) [inline mango-cli/src/main/scala/org/bdgenomics/mango/cli/VizReads.scala]: so future get requests return faster

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala (hunk 1 above)
Comment (erictu) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: this is so future get requests return quickly

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 6]
```diff
@@ -59,39 +59,23 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
...
     if (dict == null) {
-      setDictionary(filePath)
+      setDictionary(seqdict)
       setPartitioner
     }
     fileMap += ((sampleId, filePath))
   }
 
-  def loadADAMSample(filePath: String): String = {
+  def loadADAMSample(filePath: String, seqDict: SequenceDictionary): String = {
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: remove parameter seqDict: SequenceDictionary and specify this in LazyMaterialization declaration. Dict should not be initially set to null.

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 7]
```diff
@@ -59,39 +59,23 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
...
-        SequenceRecord("chr3", 2000L)))
-    }
+  def setDictionary(seqDict: SequenceDictionary) {
+    dict = seqDict
   }
 
   // Stores location of sample at a given filepath
-  def loadSample(sampleId: String, filePath: String) {
+  def loadSample(sampleId: String, filePath: String, seqdict: SequenceDictionary) {
     if (dict == null) {
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: again, remove parameter seqDict: SequenceDictionary and specify this in LazyMaterialization declaration. Dict should not be initially set to null.

File: mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala [hunk 8]
```diff
@@ -59,39 +59,23 @@ class LazyMaterialization[T: ClassTag](sc: SparkContext, partitions: Int, chunkS
...
+      setDictionary(seqdict)
       setPartitioner
     }
     fileMap += ((sampleId, filePath))
   }
 
-  def loadADAMSample(filePath: String): String = {
+  def loadADAMSample(filePath: String, seqDict: SequenceDictionary): String = {
     val region = ReferenceRegion("new", 0, chunkSize - 1)
+    setDictionary(seqDict)
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/models/LazyMaterialization.scala]: remove function setDictionary(seqDict). Like stated previously, this should be set on initialization

**Pull request merged**